### PR TITLE
Show why configuring dependencies failed

### PR DESCRIPTION
**Pull request by 9999years**
When `act-as-setup configure` fails, it doesn't indicate which dependencies are missing entirely and which ones don't satisfy the version constraints. The errors from `cabal build` are more informative, but aren't available in all contexts (nixpkgs Haskell builds, for example, use the `act-as-setup` interface). This makes it immediately clear what sort of dependency error has occurred.

Before (cabal-install 3.12.1.0):

    Configuring test-pkg-0.1.0.0...
    Error: [Cabal-8010]
    Encountered missing or private dependencies:
        base <=4.18,
        foobar,
        test-pkg:{bar-internal, foo-internal}

After:

    Configuring test-pkg-0.1.0.0...
    Error: [Cabal-8010]
    Encountered missing or private dependencies:
        base <=4.18 (installed: 4.19.1.0),
        foobar (missing),
        test-pkg:{bar-internal,foo-internal} (missing: :bar-internal)

Please let me know if you're interested in this patch and I can write tests and polish up the documentation.

---

Please read [Github PR Conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#github-pull-request-conventions) and then fill in *one* of these two templates.

---

**Template Α: This PR modifies [behaviour or interface](https://github.com/cabalism/cabal/blob/master/CONTRIBUTING.md#changelog)**

Include the following checklist in your PR:

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.
* [ ] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [ ] Tests have been added. (*Ask for help if you don’t know how to write them! Ask for an exemption if tests are too complex for too little coverage!*)

### CODE REVIEW

File: Cabal/src/Distribution/Simple/PackageIndex.hs [hunk 1]
```diff
@@ -509,23 +525,48 @@ lookupInternalDependency
...
+
+
+-- | Filter a set of installed packages to ones elligible as dependencies.
+--
+-- When we select for dependencies, we ONLY want to pick up indefinite
+-- packages, or packages with no instantiations.  We'll do mix-in linking to
+-- improve any such package into an instantiated one later.
+--
+-- INVARIANT: List of eligible 'IPI.InstalledPackageInfo' is non-empty.
+elligibleDependencies
```
Comment (geekosaur) [inline Cabal/src/Distribution/Simple/PackageIndex.hs:539]:
```suggestion
eligibleDependencies
```

File: Cabal/src/Distribution/Simple/PackageIndex.hs [hunk 2]
```diff
@@ -509,23 +525,48 @@ lookupInternalDependency
...
+  :: [(Version, [IPI.InstalledPackageInfo])]
+  -> [(Version, [IPI.InstalledPackageInfo])]
+elligibleDependencies versions =
```
Comment (geekosaur) [inline Cabal/src/Distribution/Simple/PackageIndex.hs:542]:
```suggestion
eligibleDependencies versions =
```

File: Cabal/src/Distribution/Simple/PackageIndex.hs [hunk 3]
```diff
@@ -509,23 +525,48 @@ lookupInternalDependency
...
+-- This can be used to filter the output of 'lookupPackageName' or
+-- 'lookupInternalPackageName'.
+--
+-- INVARIANT: List of eligible 'IPI.InstalledPackageInfo' is non-empty.
+matchingDependencies
+  :: VersionRange
+  -> [(Version, [IPI.InstalledPackageInfo])]
+  -> [(Version, [IPI.InstalledPackageInfo])]
+matchingDependencies versionRange versions =
+  let elligibleVersions = elligibleDependencies versions
```
Comment (geekosaur) [inline Cabal/src/Distribution/Simple/PackageIndex.hs:564]:
```suggestion
  let elligibleVersions = eligibleDependencies versions
```

File: Cabal/src/Distribution/Simple/PackageIndex.hs [hunk 4]
```diff
@@ -509,23 +525,48 @@ lookupInternalDependency
...
+   in [ (ver, pkgs)
+      | (ver, pkgs) <- elligibleVersions
```
Comment (geekosaur) [inline Cabal/src/Distribution/Simple/PackageIndex.hs:566]:
```suggestion
      | (ver, pkgs) <- eligibleVersions
```

File: Cabal/src/Distribution/Simple/PackageIndex.hs [hunk 5]
```diff
@@ -93,6 +94,10 @@ module Distribution.Simple.PackageIndex
   , dependencyCycles
   , dependencyGraph
   , moduleNameIndex
+
+    -- ** Filters on lookup results
+  , elligibleDependencies
```
Comment (geekosaur) [inline Cabal/src/Distribution/Simple/PackageIndex.hs:99]:
```suggestion
  , eligibleDependencies
```

File: Cabal/src/Distribution/Simple/Configure.hs [hunk 6]
```diff
@@ -1532,12 +1537,31 @@ dependencySatisfiable
       isInternalDep = pn == depName
 
       depSatisfiable =
-        not . null $ PackageIndex.lookupDependency installedPackageSet depName vr
+         let allVersions = PackageIndex.lookupPackageName installedPackageSet depName
+             elligibleVersions =
```
Comment (geekosaur) [inline Cabal/src/Distribution/Simple/Configure.hs:1541]:
```suggestion
             eligibleVersions =
```

File: Cabal/src/Distribution/Simple/Configure.hs [hunk 7]
```diff
@@ -1532,12 +1537,31 @@ dependencySatisfiable
...
+               [ version
+               | (version, _infos) <- PackageIndex.elligibleDependencies allVersions
```
Comment (geekosaur) [inline Cabal/src/Distribution/Simple/Configure.hs:1543]:
```suggestion
               | (version, _infos) <- PackageIndex.eligibleDependencies allVersions
```

File: Cabal/src/Distribution/Simple/Configure.hs [hunk 8]
```diff
@@ -1532,12 +1537,31 @@ dependencySatisfiable
...
+               ]
+          in if null $ PackageIndex.matchingDependencies vr allVersions
+                then
+                  if null elligibleVersions
```
Comment (geekosaur) [inline Cabal/src/Distribution/Simple/Configure.hs:1547]:
```suggestion
                  if null eligibleVersions
```

File: Cabal/src/Distribution/Simple/Configure.hs [hunk 9]
```diff
@@ -1532,12 +1537,31 @@ dependencySatisfiable
...
+                     then Unsatisfied $ MissingPackage
+                     else Unsatisfied $ WrongVersion elligibleVersions
```
Comment (geekosaur) [inline Cabal/src/Distribution/Simple/Configure.hs:1549]:
```suggestion
                     else Unsatisfied $ WrongVersion eligibleVersions
```

**Review by geekosaur — APPROVED** [review]
Unfortunately, I'm not qualified to determine whether this is logically correct, only that it looks reasonable (and being a spell checker 😛).

**Comment by 9999years** [comment]
Ah, the tests are failing because they pass the exact ID of the `base` library (which varies per GHC version under test), and that shows up in the output. Not sure how to work around that with the expect tests.

**Comment by jasagredo** [comment]
You should probably work out how to identify those and replace them with a common string in OutputNormalizer.hs

**Comment by 9999years** [comment]
@jasagredo Fixed the tests and added a new one to demonstrate the error messages I showed off in the PR description, PTAL!

File: Cabal-syntax/src/Distribution/Types/LibraryName.hs [hunk 10]
```diff
@@ -10,13 +10,15 @@ module Distribution.Types.LibraryName
...
     -- * Pretty & Parse
+  , prettyLibraryNames
   , prettyLibraryNameComponent
   , parsecLibraryNameComponent
   ) where
 
 import Distribution.Compat.Prelude
 import Prelude ()
 
+import qualified Data.List.NonEmpty as NEL
```
Comment (jasagredo) [inline Cabal-syntax/src/Distribution/Types/LibraryName.hs:21]: I don't really know the grouping followed in the cabal codebase but it feels like a block where all are `Distribution.` should not have `Data.List.NonEmpty` in it.

File: Cabal-syntax/src/Distribution/Types/DependencySatisfaction.hs
Comment (jasagredo) [inline Cabal-syntax/src/Distribution/Types/DependencySatisfaction.hs:1]: These files seem like they could be unified into `Dependency.hs`, don't they? They are very small and don't have companion functions (except for a pprinter).

File: Cabal-syntax/src/Distribution/Pretty.hs [hunk 11]
```diff
@@ -10,6 +10,8 @@ module Distribution.Pretty
   , showTokenStr
   , showFreeText
   , showFreeTextV3
+  , commaSpaceSep
```
Comment (jasagredo) [inline Cabal-syntax/src/Distribution/Pretty.hs:13]: This should not pass formatting, or I am wrong in my understanding of fourmolu. Doesn't the exported items get sorted alphabetically?

**Review by jasagredo — COMMENTED** [review]
I think this looks good in general :+1:. I am not very knowledgeable in dependency resolution, so I cannot tell if there are other places that would need changes, but the changes that were done here look correct. Basically propagating some more information.

I added some comments, in particular I would merge those 3 new modules into `Dependency.hs`, but that I guess is personal preference.

File: Cabal-syntax/src/Distribution/Pretty.hs (hunk 11 above)
Comment (geekosaur) [reply Cabal-syntax/src/Distribution/Pretty.hs:13]: No. The export list also controls the ordering and grouping of the generated haddock.

Comment (jasagredo) [reply Cabal-syntax/src/Distribution/Pretty.hs:13]: Ah my bad then. Stylish-Haskell does reorder the export list by default

File: Cabal-syntax/src/Distribution/Types/DependencySatisfaction.hs
Comment (9999years) [reply Cabal-syntax/src/Distribution/Types/DependencySatisfaction.hs:1]: I like to keep modules as small as possible. It helps keep compile times down (modules can be compiled in parallel) and helps prevent modules from growing into the 1000+-line behemoths Cabal has many of.

**Comment by jasagredo** [comment]
Aaaarg so sorry @9999years I accidentally pressed `Update branch` while scrolling. Let me try to re-push your previous version.

**Comment by jasagredo** [comment]
Restored :+1:

**Review by jasagredo — APPROVED** [review]

**Review by fgaz — DISMISSED** [review]
This is missing a changelog entry

**Comment by 9999years** [comment]
@fgaz Added a changelog entry, PTAL.

**Comment by geekosaur** [comment]
This is an API change, so can't be backported to 3.12. Additionally, 3.10 is now frozen (largely because major merge conflicts will happen thanks to fourmolu) so no backports can be done to it.

**Pull request merged**